helpers/remoteoauth/token.go

```diff
@@ -0,0 +1,165 @@
+package remoteoauth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	cloudquery_api "github.com/cloudquery/cloudquery-api-go"
+	"github.com/google/uuid"
+	"golang.org/x/oauth2"
+)
+
+type Token struct {
+	AccessToken string
+	TokenType   string
+	Expiry      time.Time
+	remoteToken bool // was the current token retrieved from a remote source
+
+	apiClient *cloudquery_api.ClientWithResponses
+	mu        sync.Mutex
+
+	cloudEnabled     bool
+	apiURL           string
+	apiToken         string
+	teamName         string
+	syncName         string
+	testConnUUID     uuid.UUID
+	syncRunUUID      uuid.UUID
+	connectorUUID    uuid.UUID
+	isTestConnection bool
+}
+
+var (
+	ErrTokenExpired = errors.New("token expired and cloud env is not set")
+
+	_ oauth2.TokenSource = (*Token)(nil)
+)
+
+// defaultExpiryDelta determines how earlier a token should be considered
+// expired than its actual expiration time. It is used to avoid late
+// expirations due to client-server time mismatches.
+const defaultExpiryDelta = 10 * time.Second
+
+func NewToken(opts ...TokenOption) (*Token, error) {
+	t := &Token{}
+	for _, opt := range opts {
+		opt(t)
+	}
+	err := t.initCloudOpts()
+	if err != nil {
+		return nil, err
+	}
+	if t.cloudEnabled {
+		t.apiClient, err = cloudquery_api.NewClientWithResponses(t.apiURL,
+			cloudquery_api.WithRequestEditorFn(func(_ context.Context, req *http.Request) error {
+				req.Header.Set("Authorization", "Bearer "+t.apiToken)
+				return nil
+			}))
+		if err != nil {
+			return nil, fmt.Errorf("failed to create api client: %w", err)
+		}
+	}
+
+	return t, nil
+}
+
+// Token returns the cached token if not expired, or a new token from the remote source.
+func (t *Token) Token() (*oauth2.Token, error) {
+	return t.TokenWithContext(context.TODO())
+}
+
+// TokenWithContext returns the cached token if not expired, or a new token from the remote source
+// using the given context.
+func (t *Token) TokenWithContext(ctx context.Context) (*oauth2.Token, error) {
+	if t.cloudEnabled && !t.remoteToken {
+		// Always retrieve token from remote source if cloud env is set
+		// and the current token is not acquired from a remote source.
+		if err := t.retrieveToken(ctx); err != nil {
+			return nil, err
+		}
+	} else if !t.Valid() {
+		if !t.cloudEnabled {
+			return nil, ErrTokenExpired
+		}
+		if err := t.retrieveToken(ctx); err != nil {
+			return nil, err
+		}
+	}
+
+	return &oauth2.Token{
+		AccessToken: t.AccessToken,
+		TokenType:   t.TokenType,
+		Expiry:      t.Expiry,
+	}, nil
+}
+
+// Valid reports whether t is non-nil, has an AccessToken, and is not expired.
+func (t *Token) Valid() bool {
```

Comment:
Don't you think isValid (i.e. using is or are for booleans) is more readable as a convention?

If not ignore the other similar comment.

Reply:
This convention is coming from golang.org/x/oauth2/oauth2.go it's not part of the TokenSource interface it's still a public method.

Reply:
TIL 👍